File: src/primitive.rs
//! Small mechanical primitives shared by profile families.
//!
//! These helpers expose spans and exact matching only. They do not decide what
//! is important, truncate arbitrary content, or infer semantics.

use crate::preservation::ByteSpan;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PrimitiveError {
    EmptyNeedle,
    EmptyPrefix,
}

/// Find exact, non-overlapping occurrences of a non-empty needle.
pub fn exact_spans(input: &str, needle: &str) -> Result<Vec<ByteSpan>, PrimitiveError> {
    if needle.is_empty() {
        return Err(PrimitiveError::EmptyNeedle);
    }

    Ok(input
        .match_indices(needle)
        .map(|(start, value)| ByteSpan::new(start, start + value.len()))
        .collect())
}

/// Find logical line-content spans whose text begins with a non-empty prefix.
///
/// Returned spans exclude LF and a directly preceding CR line terminator.
pub fn line_prefix_spans(input: &str, prefix: &str) -> Result<Vec<ByteSpan>, PrimitiveError> {
    if prefix.is_empty() {
        return Err(PrimitiveError::EmptyPrefix);
    }

    let mut spans = Vec::new();
    let mut offset = 0;

    for chunk in input.split_inclusive('\n') {
        let content_len = line_content_len(chunk);
        let content = &chunk[..content_len];

        if content.starts_with(prefix) && !content.is_empty() {
            spans.push(ByteSpan::new(offset, offset + content_len));
        }

        offset += chunk.len();
    }

    if offset < input.len() {
        let tail = &input[offset..];
        if tail.starts_with(prefix) && !tail.is_empty() {
            spans.push(ByteSpan::new(offset, input.len()));
        }
    }

    Ok(spans)
}

/// Find logical line-content spans that exactly equal a non-empty value.
pub fn exact_line_spans(input: &str, value: &str) -> Result<Vec<ByteSpan>, PrimitiveError> {
    if value.is_empty() {
        return Err(PrimitiveError::EmptyNeedle);
    }

    let mut spans = Vec::new();
    let mut offset = 0;

    for chunk in input.split_inclusive('\n') {
        let content_len = line_content_len(chunk);
        let content = &chunk[..content_len];

        if content == value {
            spans.push(ByteSpan::new(offset, offset + content_len));
        }

        offset += chunk.len();
    }

    if offset < input.len() && &input[offset..] == value {
        spans.push(ByteSpan::new(offset, input.len()));
    }

    Ok(spans)
}

fn line_content_len(chunk: &str) -> usize {
    let without_lf = chunk.strip_suffix('\n').unwrap_or(chunk);
    without_lf
        .strip_suffix('\r')
        .map_or(without_lf.len(), str::len)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn exact_spans_are_ordered_and_non_overlapping() {
        let spans = exact_spans("aa xx aa aa", "aa").unwrap();
        assert_eq!(
            spans,
            vec![
                ByteSpan::new(0, 2),
                ByteSpan::new(6, 8),
                ByteSpan::new(9, 11)
            ]
        );
    }

    #[test]
    fn exact_spans_respect_utf8_boundaries() {
        let input = "é中é";
        let spans = exact_spans(input, "é").unwrap();
        assert_eq!(spans, vec![ByteSpan::new(0, 2), ByteSpan::new(5, 7)]);
        for span in spans {
            assert!(span.validate(input).is_ok());
        }
    }

    #[test]
    fn line_helpers_handle_lf_crlf_and_tail() {
        let input = "FAIL one\r\nok\nFAIL two";
        assert_eq!(
            line_prefix_spans(input, "FAIL").unwrap(),
            vec![ByteSpan::new(0, 8), ByteSpan::new(12, 20)]
        );
        assert_eq!(
            exact_line_spans(input, "ok").unwrap(),
            vec![ByteSpan::new(10, 12)]
        );
    }

    #[test]
    fn empty_matching_terms_are_rejected() {
        assert_eq!(exact_spans("x", ""), Err(PrimitiveError::EmptyNeedle));
        assert_eq!(line_prefix_spans("x", ""), Err(PrimitiveError::EmptyPrefix));
        assert_eq!(exact_line_spans("x", ""), Err(PrimitiveError::EmptyNeedle));
    }
}

File: src/primitive.ts
import type { ByteSpan } from "./preservation.js";

export class PrimitiveError extends Error {
  override readonly name = "PrimitiveError";
}

export function exactSpans(input: string, needle: string): ByteSpan[] {
  if (needle.length === 0) {
    throw new PrimitiveError("needle must not be empty");
  }

  const spans: ByteSpan[] = [];
  let from = 0;
  while (from <= input.length - needle.length) {
    const start = input.indexOf(needle, from);
    if (start < 0) {
      break;
    }
    spans.push({ start_byte: utf8Offset(input, start), end_byte: utf8Offset(input, start + needle.length) });
    from = start + needle.length;
  }
  return spans;
}

export function linePrefixSpans(input: string, prefix: string): ByteSpan[] {
  if (prefix.length === 0) {
    throw new PrimitiveError("prefix must not be empty");
  }

  return lineSpans(input, (line) => line.startsWith(prefix));
}

export function exactLineSpans(input: string, value: string): ByteSpan[] {
  if (value.length === 0) {
    throw new PrimitiveError("value must not be empty");
  }

  return lineSpans(input, (line) => line === value);
}

function lineSpans(
  input: string,
  predicate: (line: string) => boolean,
): ByteSpan[] {
  const spans: ByteSpan[] = [];
  let charOffset = 0;

  for (const chunk of splitInclusiveLf(input)) {
    const content = lineContent(chunk);
    if (predicate(content)) {
      spans.push({
        start_byte: utf8Offset(input, charOffset),
        end_byte: utf8Offset(input, charOffset + content.length),
      });
    }
    charOffset += chunk.length;
  }

  return spans;
}

function lineContent(chunk: string): string {
  if (!chunk.endsWith("\n")) {
    return chunk;
  }

  const withoutLf = chunk.slice(0, -1);
  return withoutLf.endsWith("\r")
    ? withoutLf.slice(0, -1)
    : withoutLf;
}

function splitInclusiveLf(input: string): string[] {
  if (input.length === 0) {
    return [];
  }

  const chunks: string[] = [];
  let start = 0;
  while (start < input.length) {
    const newline = input.indexOf("\n", start);
    if (newline < 0) {
      chunks.push(input.slice(start));
      break;
    }
    chunks.push(input.slice(start, newline + 1));
    start = newline + 1;
  }
  return chunks;
}

function utf8Offset(input: string, charOffset: number): number {
  return Buffer.byteLength(input.slice(0, charOffset), "utf8");
}
